core/src/main/java/org/mule/runtime/core/internal/context/thread/notification/DefaultThreadNotificationService.java

```diff
@@ -0,0 +1,134 @@
+ *
+ * @since 4.2
+ */
+public class DefaultThreadNotificationService implements ThreadNotificationService, Lifecycle {
+
+  private static final int DEFAULT_INTERVAL = 5;
+  private static final String newLine = System.getProperty("line.separator");
+  private ThreadsStatistics stats = new ThreadsStatistics();
+  private int interval;
+
+  @Inject
+  private SchedulerService schedulerService;
+
+  private Scheduler scheduler;
+
+  public DefaultThreadNotificationService() {
+    this(DEFAULT_INTERVAL);
+  }
+
+  public DefaultThreadNotificationService(int interval) {
+    this.interval = interval;
+  }
+
+  @Override
+  public void addThreadNotificationElement(ThreadNotificationElement notification) {
+    stats.addThreadNotificationElement(notification);
+  }
+
+  @Override
+  public void addThreadNotificationElements(Collection<ThreadNotificationElement> notifications) {
+    stats.addThreadNotificationElements(notifications);
+  }
+
+  @Override
+  public String getNotification() {
+    String msg = "";
+    for (Pair<String, String> transition : stats.getPossibleTransitions()) {
+      msg += getNotificationFor(transition);
+    }
+    return msg;
+  }
+
+  private String getNotificationFor(Pair<String, String> transition) {
+    String count = formattedNumber(stats.getCount(transition));
+    String mean = formattedNumber(stats.getMean(transition));
+    String std = formattedNumber(stats.getStdDeviation(transition));
+    String percentile = formattedNumber(stats.percentile(transition, 0.9));
+
+    Function<String, String> pad = getPad(count, mean, std, percentile);
+
+    String msg = newLine + "Stats for transition " + transition.getFirst() + " - " + transition.getSecond() + newLine;
+    msg += " Transitions count: " + pad.apply(count) + newLine;
+    msg += "              Mean: " + pad.apply(mean) + " nSecs" + newLine;
+    msg += "Standard Deviation: " + pad.apply(std) + " nSecs" + newLine;
+    msg += "    Percentile 90%: " + pad.apply(percentile) + " nSecs" + newLine;
+    return msg;
+  }
+
+  @Override
+  public void clear() {
+    stats.clear();
+  }
+
+  private void logStats() {
+    for (Pair<String, String> transition : stats.getPossibleTransitions()) {
+      // We set this to warn logging level because the entire pipeline is already created, so even if the
+      // logging level changes, the pipeline should be defined using a system property (the ThreadNotificationLogger
+      // needs to know whether to add logging phases or not).
+      REPORT_LOGGER.warn(getNotificationFor(transition));
+    }
+  }
+
+  private Function<String, String> getPad(String... elements) {
+    final Integer size = Arrays.stream(elements).map(element -> element.length()).max(Integer::compare).get();
+    return value -> leftPad(value, size);
+  }
+
+  private String formattedNumber(double nanos) {
+    return format("%,d", new Double(nanos).intValue());
+  }
+
+  @Override
+  public void dispose() {
```

Comment:
why not implement just Startable and Stoppable instead of Lifecycle and have this empty methods?